dvc/repo/diff.py

```diff
@@ -1,246 +1,97 @@
             )
-        if not diff_dct[DIFF_DELETED]:
-            result[DIFF_NEW_FILE] = target
-            result[DIFF_NEW_CHECKSUM] = diff_dct[DIFF_B_OUTPUT].checksum
-            size += os.path.getsize(
-                self.cache.local.get(diff_dct[DIFF_B_OUTPUT].checksum)
+        ] + [
+            Diffable(
+                filename=str(output.path_info / entry["relpath"]),
+                checksum=entry["md5"],
             )
-        result[DIFF_SIZE] = size
-    except IOError as e:
-        _file_not_exists(e, result)
-    return result
-
+            for entry in output.dir_cache
+        ]
 
-def _diff_royal(self, target, diff_dct):
-    if diff_dct[DIFF_IS_DIR]:
-        return _diff_dir(self, target, diff_dct)
-    return _diff_file(self, target, diff_dct)
+    return [Diffable(filename=str(output), checksum=output.checksum)]
 
 
 @locked
-def diff(self, a_ref, target=None, b_ref=None):
-    """Generates diff message string output
+def diff(self, a_ref="HEAD", b_ref=None, *, target=None):
+    """
+    By default, it compares the working tree with the last commit's tree.
 
-    Args:
-        target(str) - file/directory to check diff of
-        a_ref(str) - first tag
-        (optional) b_ref(str) - second git tag
+    When a `target` path is given, it only shows that file's comparison.
 
-    Returns:
-        string: string of output message with diff info
+    This implementation differs from `git diff` since DVC doesn't have
+    the concept of `index`, but it keeps the same interface, thus,
+    `dvc diff` would be the same as `dvc diff HEAD`.
     """
-    result = {}
-    diff_dct = self.scm.get_diff_trees(a_ref, b_ref=b_ref)
-    result[DIFF_A_REF] = diff_dct[DIFF_A_REF]
-    result[DIFF_B_REF] = diff_dct[DIFF_B_REF]
-    if diff_dct[DIFF_EQUAL]:
-        result[DIFF_EQUAL] = True
-        return result
-    result[DIFF_LIST] = []
-    diff_outs = _get_diff_outs(self, diff_dct)
-    if target is None:
-        result[DIFF_LIST] = [
-            _diff_royal(self, path, diff_outs[path]) for path in diff_outs
-        ]
-    elif target in diff_outs:
-        result[DIFF_LIST] = [_diff_royal(self, target, diff_outs[target])]
-    else:
-        msg = "Have not found file/directory '{}' in the commits"
-        raise FileNotInCommitError(msg.format(target))
+    if type(self.scm) is not Git:
+        raise DvcException("only supported for Git repositories")
+
+    outs = {}
+
+    for branch in self.brancher(revs=[a_ref, b_ref]):
+        outs[branch] = set(
+            diffable
+            for stage in self.stages
+            for out in stage.outs
+            for diffable in _diffables_from_output(out)
+            if not target or target == str(out)
+        )
+
+    old = outs[a_ref]
+    new = outs[b_ref or "working tree"]
+
+    added = new - old
+    deleted = old - new
+    delta = old ^ new
+
+    result = {
+        "added": [entry._asdict() for entry in sorted(added)],
```

Comment:
Converting the entry to dict to keep a consistent output. However, I could return Diffable entries